libraries/anvill_passes/src/RecoverStackFrameInformation.cpp

```diff
@@ -0,0 +1,437 @@
+RecoverStackFrameInformation::GenerateStackFrameType(
+    const llvm::Function &function,
+    const StackFrameAnalysis &stack_frame_analysis, std::size_t padding_bytes) {
+  if (stack_frame_analysis.instruction_list.empty()) {
+    return StackAnalysisErrorCode::InvalidParameter;
+  }
+
+  // Generate a stack frame type with a name that matches the anvill ABI
+  auto function_name = function.getName().str();
+  auto stack_frame_type_name = function_name + kStackFrameTypeNameSuffix;
+
+  // Make sure this type is not defined already
+  auto &module = *function.getParent();
+  auto stack_frame_type = module.getTypeByName(stack_frame_type_name);
+  if (stack_frame_type != nullptr) {
+    return StackAnalysisErrorCode::StackFrameTypeAlreadyExists;
+  }
+
+  // Determine how many bytes we should allocate. We may have been
+  // asked to add some additional padding. We don't care how it is
+  // accessed right now, we just add to the total size of the final
+  // stack frame
+  auto stack_frame_size = padding_bytes + stack_frame_analysis.size;
+
+  // Generate the stack frame using a byte array
+  auto &context = module.getContext();
+
+  auto array_elem_type = llvm::Type::getInt8Ty(context);
+  auto byte_array_type =
+      llvm::ArrayType::get(array_elem_type, stack_frame_size);
+
+  const std::vector<llvm::Type *> stack_frame_types = {byte_array_type};
+  stack_frame_type =
+      llvm::StructType::create(stack_frame_types, stack_frame_type_name, true);
+
+  if (stack_frame_type == nullptr) {
+    return StackAnalysisErrorCode::InternalError;
+  }
+
+  return stack_frame_type;
+}
+
+Result<llvm::GlobalVariable *, StackAnalysisErrorCode>
+RecoverStackFrameInformation::GetStackSymbolicByteValue(llvm::Module &module,
+                                                        std::int32_t offset) {
+  // Create a new name
+  auto value_name = kSymbolicStackFrameValuePrefix;
+  if (offset < 0) {
+    value_name += "minus_";
+  } else if (offset > 0) {
+    value_name += "plus_";
+  }
+
+  value_name += std::to_string(abs(offset));
+
+  // Get the symbolic value; this will fail if we already have
+  // one with the same name but wrong type
+  auto &context = module.getContext();
+  auto byte_type = llvm::Type::getInt8Ty(context);
+
+  auto symbolic_value_res = GetSymbolicValue(module, byte_type, value_name);
+  if (!symbolic_value_res.Succeeded()) {
+    return StackAnalysisErrorCode::StackInitializationError;
+  }
+
+  return symbolic_value_res.TakeValue();
+}
+
+Result<std::monostate, StackAnalysisErrorCode>
+RecoverStackFrameInformation::UpdateFunction(
+    llvm::Function &function, const StackFrameAnalysis &stack_frame_analysis,
+    StackFrameStructureInitializationProcedure init_strategy,
+    std::size_t stack_frame_lower_padding,
+    std::size_t stack_frame_higher_padding) {
+
+  if (function.isDeclaration() ||
+      stack_frame_analysis.instruction_list.empty()) {
+    return StackAnalysisErrorCode::InvalidParameter;
```

Comment:
Here `instruction_list` could be empty for some function and it should not throw an error.  e.g:
```
define i32 @foo() #0 {
  %1 = call i32 @bar()
  ret i32 %1
}
```

Reply:
Can you check if this shows up in Carson's branch?